options/ansi/generic/time-stubs.cpp

```diff
@@ -1089,7 +1092,155 @@ time_t timelocal(struct tm *) {
+	DECL_LEAP_SECOND(1981, 1, 0),
+	DECL_LEAP_SECOND(1982, 1, 0),
+	DECL_LEAP_SECOND(1983, 1, 0),
+	DECL_LEAP_SECOND(1984, 0, 0),
+	DECL_LEAP_SECOND(1985, 1, 0),
+	DECL_LEAP_SECOND(1986, 0, 0),
+	DECL_LEAP_SECOND(1987, 0, 1),
+	DECL_LEAP_SECOND(1988, 0, 0),
+	DECL_LEAP_SECOND(1989, 0, 1),
+	DECL_LEAP_SECOND(1990, 0, 1),
+	DECL_LEAP_SECOND(1991, 0, 0),
+	DECL_LEAP_SECOND(1992, 1, 0),
+	DECL_LEAP_SECOND(1993, 1, 0),
+	DECL_LEAP_SECOND(1994, 1, 0),
+	DECL_LEAP_SECOND(1995, 0, 1),
+	DECL_LEAP_SECOND(1996, 0, 0),
+	DECL_LEAP_SECOND(1997, 1, 0),
+	DECL_LEAP_SECOND(1998, 0, 1),
+	DECL_LEAP_SECOND(1999, 0, 0),
+	DECL_LEAP_SECOND(2000, 0, 0),
+	DECL_LEAP_SECOND(2001, 0, 0),
+	DECL_LEAP_SECOND(2002, 0, 0),
+	DECL_LEAP_SECOND(2003, 0, 0),
+	DECL_LEAP_SECOND(2004, 0, 0),
+	DECL_LEAP_SECOND(2005, 0, 1),
+	DECL_LEAP_SECOND(2006, 0, 0),
+	DECL_LEAP_SECOND(2007, 0, 0),
+	DECL_LEAP_SECOND(2008, 0, 1),
+	DECL_LEAP_SECOND(2009, 0, 0),
+	DECL_LEAP_SECOND(2010, 0, 0),
+	DECL_LEAP_SECOND(2011, 0, 0),
+	DECL_LEAP_SECOND(2012, 1, 0),
+	DECL_LEAP_SECOND(2013, 0, 0),
+	DECL_LEAP_SECOND(2014, 0, 0),
+	DECL_LEAP_SECOND(2015, 1, 0),
+	DECL_LEAP_SECOND(2016, 0, 1),
+};
+
+static time_t get_leap_second(int year, int month) {
+	const time_t num_years = sizeof(leap_seconds) / sizeof(leap_seconds[0]);
+	if(year < 1970) {
+		return 0;
+	}
+	if(num_years <= year-1970) {
+		return 0;
+	}
+	return leap_seconds[year - 1970][month];
+}
+
+static time_t leap_seconds_in_year(int year) {
+	time_t ret = 0;
+	for(int i = 0; i < 12; i++) {
+		ret += get_leap_second(year, i);
+	}
+	return ret;
+}
+
+static bool is_leap_year(int year) {
+	return (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
+}
+
+static time_t days_in_year(int year) {
+	return DAYS_JANUARY +
+	       DAYS_FEBRUARY + (is_leap_year(year) ? 1 : 0) +
+	       DAYS_MARCH +
+	       DAYS_APRIL +
+	       DAYS_MAY +
+	       DAYS_JUNE +
+	       DAYS_JULY +
+	       DAYS_AUGUST +
+	       DAYS_SEPTEMBER +
+	       DAYS_OCTOBER +
+	       DAYS_NOVEMBER +
+	       DAYS_DECEMBER;
+}
+
+time_t timegm(struct tm *tm) {
+	time_t year = tm->tm_year + 1900;
+	time_t month = tm->tm_mon;
+	time_t day = tm->tm_mday - 1;
+	time_t hour = tm->tm_hour;
+	time_t minute = tm->tm_min;
+	time_t second = tm->tm_sec;
+
+	time_t ret = 0;
+	for(time_t y = 1970; y < year; y++) {
+		time_t year_leaps = leap_seconds_in_year(y);
+		time_t year_days = days_in_year(y);
+		time_t year_seconds = year_days * 24 * 60 * 60 + year_leaps;
+		ret += year_seconds;
+	}
+
+	int month_days_list[12] = {
+		DAYS_JANUARY,
+		DAYS_FEBRUARY + (is_leap_year(year) ? 1 : 0),
+		DAYS_MARCH,
+		DAYS_APRIL,
+		DAYS_MAY,
+		DAYS_JUNE,
+		DAYS_JULY,
+		DAYS_AUGUST,
+		DAYS_SEPTEMBER,
+		DAYS_OCTOBER,
+		DAYS_NOVEMBER,
+		DAYS_DECEMBER,
+	};
+
+	for(uint8_t m = 0; m < month; m++) {
+		int month_leaps = get_leap_second(year, m);
+		int month_days = month_days_list[m];
+		int month_seconds = month_days * 24 * 60 * 60 + month_leaps;
+		ret += month_seconds;
+	}
```

Comment:
further up time_t was assigned to directly, cant it be used here to avoid the possibly narrowing implicit cast from int